// my-ui-lib/index.ts
import {NgModule, ModuleWithProviders} from '@angular/core';
import {CommonModule} from '@angular/common';
import {EditableTableComponent} from './editable-table/editable-table.component';
import {TableComponentComponent} from './table-component/table-component.component';
import {ObjectKeysPipe} from "./object-keys.pipe";
import {DatePickerComponent} from "./date-picker/date-picker.component";

export * from './editable-table/editable-table.component';
export * from './table-component/table-component.component';
export * from './object-keys.pipe';
export * from './shared/displayedObject';
export * from './shared/editableObject';
export * from './shared/editedObject';
export * from './shared/tableCell';
export * from './shared/editableTableCell';
export * from './shared/tableRow';
export * from './date-picker/date-picker.component';

@NgModule({
    imports: [
        CommonModule
    ],
    declarations: [
        EditableTableComponent,
        TableComponentComponent,
        DatePickerComponent,
        ObjectKeysPipe
    ],
    exports: [
        EditableTableComponent,
        TableComponentComponent,
        DatePickerComponent,
        ObjectKeysPipe
    ]
})
export class TableModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: TableModule,
            providers: []
        };
    }
}
